File: sdks/ios/Sources/OpenEventFlow/OpenEventFlowClient.swift
import Foundation

public final class OpenEventFlowClient: Analytics {
    private let app: AppContext
    private let transport: AnalyticsTransport
    private let store: EventStore
    private let idGenerator: () -> String
    private let clock: () -> Date
    private var consent = AnalyticsConsent(analyticsAllowed: true)
    private var anonymousId: String
    private var userId: String?
    private var traits: [String: Any] = [:]
    private var session: SessionContext
    private var stays: [String: StayState] = [:]
    private let minStayDurationMs: Int64 = 1_000
    private let maxStayDurationMs: Int64 = 24 * 60 * 60 * 1_000

    public init(
        app: AppContext,
        transport: AnalyticsTransport,
        store: EventStore = MemoryEventStore(),
        idGenerator: @escaping () -> String = { UUID().uuidString },
        clock: @escaping () -> Date = Date.init
    ) {
        self.app = app
        self.transport = transport
        self.store = store
        self.idGenerator = idGenerator
        self.clock = clock
        self.anonymousId = idGenerator()
        self.session = SessionContext(sessionId: idGenerator(), startedAt: clock())
    }

    public func track(_ event: AnalyticsEvent) {
        guard consent.analyticsAllowed else { return }
        store.push(
            NormalizedEvent(
                eventId: idGenerator(),
                eventName: event.name,
                schema: event.schema,
                clientTime: clock(),
                properties: sanitize(event.properties),
                context: EventContext(
                    app: app,
                    user: UserContext(
                        anonymousId: anonymousId,
                        userId: userId,
                        traits: consent.piiAllowed ? traits : [:]
                    ),
                    session: session,
                    consent: consent
                )
            )
        )
    }

    public func screen(_ name: String, properties: [String: Any] = [:]) {
        var merged = properties
        merged["name"] = name
        track(
            SimpleAnalyticsEvent(
                name: "screen_view",
                schema: "iglu:io.openeventflow/screen_view/jsonschema/1-0-0",
                properties: merged
            )
        )
    }

    public func identify(userId: String, traits: [String: Any] = [:]) {
        self.userId = userId
        self.traits = traits
    }

    public func setConsent(_ consent: AnalyticsConsent) {
        self.consent = consent
    }

    public func flush() {
        let batch = store.peek(limit: 50)
        guard !batch.isEmpty else { return }
        transport.send(batch)
        store.remove(count: batch.count)
    }

    @discardableResult
    public func startNewSession() -> String {
        session = SessionContext(sessionId: idGenerator(), startedAt: clock())
        return session.sessionId
    }

    public func queueSize() -> Int {
        store.size()
    }

    @discardableResult
    public func beginStay(key: String, properties: [String: Any] = [:]) -> StayStartResult {
        precondition(!key.isEmpty, "stay key is required")
        let startedAt = clock()
        let stay = StayState(
            key: key,
            stayId: idGenerator(),
            properties: properties,
            startedAt: startedAt,
            activeStartedAt: startedAt
        )
        stays[key] = stay
        return StayStartResult(accepted: true, stayId: stay.stayId, startedAt: startedAt)
    }

    @discardableResult
    public func pauseStay(key: String) -> StayUpdateResult {
        guard var stay = stays[key] else {
            return StayUpdateResult(accepted: false, reason: "stay_not_found")
        }
        if stay.paused {
            return StayUpdateResult(accepted: true, durationMs: currentStayDuration(stay, now: clock()))
        }
        let now = clock()
        stay.accumulatedMs += millisBetween(stay.activeStartedAt, now)
        stay.paused = true
        stay.pausedAt = now
        stays[key] = stay
        return StayUpdateResult(accepted: true, durationMs: clampStayDuration(stay.accumulatedMs))
    }

    @discardableResult
    public func resumeStay(key: String) -> StayUpdateResult {
        guard var stay = stays[key] else {
            return StayUpdateResult(accepted: false, reason: "stay_not_found")
        }
        if !stay.paused {
            return StayUpdateResult(accepted: true, durationMs: currentStayDuration(stay, now: clock()))
        }
        stay.activeStartedAt = clock()
        stay.paused = false
        stay.pausedAt = nil
        stays[key] = stay
        return StayUpdateResult(accepted: true, durationMs: clampStayDuration(stay.accumulatedMs))
    }

    @discardableResult
    public func endStay(
        key: String,
        exitReason: String = "unknown",
        properties: [String: Any] = [:]
    ) -> StayEndResult {
        guard let stay = stays.removeValue(forKey: key) else {
            return StayEndResult(accepted: false, reason: "stay_not_found")
        }
        let durationMs = currentStayDuration(stay, now: clock())
        guard durationMs >= minStayDurationMs else {
            return StayEndResult(
                accepted: false,
                durationMs: durationMs,
                stayId: stay.stayId,
                reason: "stay_duration_below_minimum"
            )
        }
        guard consent.analyticsAllowed else {
            return StayEndResult(
                accepted: false,
                durationMs: durationMs,
                stayId: stay.stayId,
                reason: "analytics_consent_disabled"
            )
        }

        var merged = stay.properties
        for (key, value) in properties {
            merged[key] = value
        }
        merged["stay_id"] = stay.stayId
        merged["duration_ms"] = durationMs
        merged["exit_reason"] = exitReason
        track(
            SimpleAnalyticsEvent(
                name: "page_stay",
                schema: "iglu:io.openeventflow/page_stay/jsonschema/1-0-0",
                properties: merged
            )
        )
        return StayEndResult(accepted: true, durationMs: durationMs, stayId: stay.stayId)
    }

    @discardableResult
    public func switchStay(
        key: String,
        properties: [String: Any] = [:],
        exitReason: String = "route_change"
    ) -> StayEndResult {
        let ended = stays[key] == nil
            ? StayEndResult(accepted: false, reason: "stay_not_found")
            : endStay(key: key, exitReason: exitReason)
        beginStay(key: key, properties: properties)
        return ended
    }

    @discardableResult
    public func cancelStay(key: String) -> StayUpdateResult {
        StayUpdateResult(accepted: stays.removeValue(forKey: key) != nil)
    }

    @discardableResult
    public func flushActiveStays(exitReason: String = "app_shutdown") -> [StayEndResult] {
        Array(stays.keys).map { key in endStay(key: key, exitReason: exitReason) }
    }

    private func sanitize(_ properties: [String: Any]) -> [String: Any] {
        guard !consent.piiAllowed else { return properties }
        return properties.filter { key, _ in
            key.range(of: "email|phone|mobile|id_card|password|address", options: [.regularExpression, .caseInsensitive]) == nil
        }
    }

    private func currentStayDuration(_ stay: StayState, now: Date) -> Int64 {
        let activeDuration = stay.paused ? 0 : millisBetween(stay.activeStartedAt, now)
        return clampStayDuration(stay.accumulatedMs + activeDuration)
    }

    private func clampStayDuration(_ durationMs: Int64) -> Int64 {
        min(max(durationMs, 0), maxStayDurationMs)
    }

    private func millisBetween(_ start: Date, _ end: Date) -> Int64 {
        Int64((end.timeIntervalSince(start) * 1000).rounded(.down))
    }
}

public protocol AnalyticsTransport {
    func send(_ batch: [NormalizedEvent])
}

public protocol EventStore {
    func push(_ event: NormalizedEvent)
    func peek(limit: Int) -> [NormalizedEvent]
    func remove(count: Int)
    func size() -> Int
}

public final class MemoryEventStore: EventStore {
    private var events: [NormalizedEvent] = []

    public init() {}

    public func push(_ event: NormalizedEvent) {
        events.append(event)
    }

    public func peek(limit: Int) -> [NormalizedEvent] {
        Array(events.prefix(limit))
    }

    public func remove(count: Int) {
        events.removeFirst(Swift.min(count, events.count))
    }

    public func size() -> Int {
        events.count
    }
}

public struct AppContext {
    public let appId: String
    public let platform: String
    public let appVersion: String
    public let sdkVersion: String

    public init(appId: String, platform: String, appVersion: String, sdkVersion: String) {
        self.appId = appId
        self.platform = platform
        self.appVersion = appVersion
        self.sdkVersion = sdkVersion
    }
}

public struct UserContext {
    public let anonymousId: String
    public let userId: String?
    public let traits: [String: Any]
}

public struct SessionContext {
    public let sessionId: String
    public let startedAt: Date
}

public struct EventContext {
    public let app: AppContext
    public let user: UserContext
    public let session: SessionContext
    public let consent: AnalyticsConsent
}

public struct NormalizedEvent {
    public let eventId: String
    public let eventName: String
    public let schema: String
    public let clientTime: Date
    public let properties: [String: Any]
    public let context: EventContext
}

public struct SimpleAnalyticsEvent: AnalyticsEvent {
    public let name: String
    public let schema: String
    public let properties: [String: Any]

    public init(name: String, schema: String, properties: [String: Any]) {
        self.name = name
        self.schema = schema
        self.properties = properties
    }
}

public struct StayStartResult {
    public let accepted: Bool
    public let stayId: String?
    public let startedAt: Date?
    public let reason: String?

    public init(accepted: Bool, stayId: String? = nil, startedAt: Date? = nil, reason: String? = nil) {
        self.accepted = accepted
        self.stayId = stayId
        self.startedAt = startedAt
        self.reason = reason
    }
}

public struct StayUpdateResult {
    public let accepted: Bool
    public let durationMs: Int64?
    public let reason: String?

    public init(accepted: Bool, durationMs: Int64? = nil, reason: String? = nil) {
        self.accepted = accepted
        self.durationMs = durationMs
        self.reason = reason
    }
}

public struct StayEndResult {
    public let accepted: Bool
    public let durationMs: Int64?
    public let stayId: String?
    public let reason: String?

    public init(accepted: Bool, durationMs: Int64? = nil, stayId: String? = nil, reason: String? = nil) {
        self.accepted = accepted
        self.durationMs = durationMs
        self.stayId = stayId
        self.reason = reason
    }
}

private struct StayState {
    let key: String
    let stayId: String
    let properties: [String: Any]
    let startedAt: Date
    var activeStartedAt: Date
    var accumulatedMs: Int64 = 0
    var paused: Bool = false
    var pausedAt: Date?
}

File: sdks/android/src/main/kotlin/io/openeventflow/mobile/OpenEventFlowClient.kt
package io.openeventflow.mobile

import java.util.UUID

class OpenEventFlowClient(
    private val app: AppContext,
    private val transport: AnalyticsTransport,
    private val store: EventStore = MemoryEventStore(),
    private val idGenerator: () -> String = { UUID.randomUUID().toString() },
    private val clock: () -> Long = { System.currentTimeMillis() }
) : Analytics {
    private var consent = AnalyticsConsent(analyticsAllowed = true)
    private var anonymousId = idGenerator()
    private var userId: String? = null
    private var traits: Map<String, Any?> = emptyMap()
    private var session = SessionContext(sessionId = idGenerator(), startedAt = clock())
    private val stays = mutableMapOf<String, StayState>()
    private val minStayDurationMs = 1_000L
    private val maxStayDurationMs = 24 * 60 * 60 * 1_000L

    override fun track(event: AnalyticsEvent) {
        if (!consent.analyticsAllowed) {
            return
        }
        store.push(
            NormalizedEvent(
                eventId = idGenerator(),
                eventName = event.name,
                schema = event.schema,
                clientTime = clock(),
                properties = sanitize(event.properties),
                context = EventContext(
                    app = app,
                    user = UserContext(
                        anonymousId = anonymousId,
                        userId = userId,
                        traits = if (consent.piiAllowed) traits else emptyMap()
                    ),
                    session = session,
                    consent = consent
                )
            )
        )
    }

    override fun screen(name: String, properties: Map<String, Any?>) {
        track(
            SimpleAnalyticsEvent(
                name = "screen_view",
                schema = "iglu:io.openeventflow/screen_view/jsonschema/1-0-0",
                properties = mapOf("name" to name) + properties
            )
        )
    }

    override fun identify(userId: String, traits: Map<String, Any?>) {
        this.userId = userId
        this.traits = traits
    }

    override fun setConsent(consent: AnalyticsConsent) {
        this.consent = consent
    }

    override fun flush() {
        val batch = store.peek(50)
        if (batch.isEmpty()) {
            return
        }
        transport.send(batch)
        store.remove(batch.size)
    }

    fun startNewSession(): String {
        session = SessionContext(sessionId = idGenerator(), startedAt = clock())
        return session.sessionId
    }

    fun queueSize(): Int = store.size()

    fun beginStay(key: String, properties: Map<String, Any?> = emptyMap()): StayStartResult {
        require(key.isNotBlank()) { "stay key is required" }
        val startedAt = clock()
        val stay = StayState(
            key = key,
            stayId = idGenerator(),
            properties = properties,
            startedAt = startedAt,
            activeStartedAt = startedAt
        )
        stays[key] = stay
        return StayStartResult(accepted = true, stayId = stay.stayId, startedAt = startedAt)
    }

    fun pauseStay(key: String): StayUpdateResult {
        val stay = stays[key] ?: return StayUpdateResult(accepted = false, reason = "stay_not_found")
        if (stay.paused) {
            return StayUpdateResult(accepted = true, durationMs = currentStayDuration(stay, clock()))
        }
        val now = clock()
        stay.accumulatedMs += now - stay.activeStartedAt
        stay.paused = true
        stay.pausedAt = now
        return StayUpdateResult(accepted = true, durationMs = clampStayDuration(stay.accumulatedMs))
    }

    fun resumeStay(key: String): StayUpdateResult {
        val stay = stays[key] ?: return StayUpdateResult(accepted = false, reason = "stay_not_found")
        if (!stay.paused) {
            return StayUpdateResult(accepted = true, durationMs = currentStayDuration(stay, clock()))
        }
        stay.activeStartedAt = clock()
        stay.paused = false
        stay.pausedAt = null
        return StayUpdateResult(accepted = true, durationMs = clampStayDuration(stay.accumulatedMs))
    }

    fun endStay(
        key: String,
        exitReason: String = "unknown",
        properties: Map<String, Any?> = emptyMap()
    ): StayEndResult {
        val stay = stays.remove(key) ?: return StayEndResult(accepted = false, reason = "stay_not_found")
        val durationMs = currentStayDuration(stay, clock())
        if (durationMs < minStayDurationMs) {
            return StayEndResult(
                accepted = false,
                durationMs = durationMs,
                stayId = stay.stayId,
                reason = "stay_duration_below_minimum"
            )
        }
        if (!consent.analyticsAllowed) {
            return StayEndResult(
                accepted = false,
                durationMs = durationMs,
                stayId = stay.stayId,
                reason = "analytics_consent_disabled"
            )
        }
        track(
            SimpleAnalyticsEvent(
                name = "page_stay",
                schema = "iglu:io.openeventflow/page_stay/jsonschema/1-0-0",
                properties = stay.properties + properties + mapOf(
                    "stay_id" to stay.stayId,
                    "duration_ms" to durationMs,
                    "exit_reason" to exitReason
                )
            )
        )
        return StayEndResult(accepted = true, durationMs = durationMs, stayId = stay.stayId)
    }

    fun switchStay(
        key: String,
        properties: Map<String, Any?> = emptyMap(),
        exitReason: String = "route_change"
    ): StayEndResult {
        val ended = if (stays.containsKey(key)) {
            endStay(key, exitReason = exitReason)
        } else {
            StayEndResult(accepted = false, reason = "stay_not_found")
        }
        beginStay(key, properties)
        return ended
    }

    fun cancelStay(key: String): StayUpdateResult {
        return StayUpdateResult(accepted = stays.remove(key) != null)
    }

    fun flushActiveStays(exitReason: String = "app_shutdown"): List<StayEndResult> {
        return stays.keys.toList().map { key -> endStay(key, exitReason = exitReason) }
    }

    private fun sanitize(properties: Map<String, Any?>): Map<String, Any?> {
        if (consent.piiAllowed) {
            return properties
        }
        return properties.filterKeys { !it.contains(Regex("email|phone|mobile|id_card|password|address", RegexOption.IGNORE_CASE)) }
    }

    private fun currentStayDuration(stay: StayState, now: Long): Long {
        val activeDuration = if (stay.paused) 0 else now - stay.activeStartedAt
        return clampStayDuration(stay.accumulatedMs + activeDuration)
    }

    private fun clampStayDuration(durationMs: Long): Long {
        return durationMs.coerceAtLeast(0).coerceAtMost(maxStayDurationMs)
    }
}

interface AnalyticsTransport {
    fun send(batch: List<NormalizedEvent>)
}

interface EventStore {
    fun push(event: NormalizedEvent)
    fun peek(limit: Int): List<NormalizedEvent>
    fun remove(count: Int)
    fun size(): Int
}

class MemoryEventStore : EventStore {
    private val events = mutableListOf<NormalizedEvent>()

    override fun push(event: NormalizedEvent) {
        events.add(event)
    }

    override fun peek(limit: Int): List<NormalizedEvent> = events.take(limit)

    override fun remove(count: Int) {
        repeat(count.coerceAtMost(events.size)) {
            events.removeAt(0)
        }
    }

    override fun size(): Int = events.size
}

data class AppContext(
    val appId: String,
    val platform: String,
    val appVersion: String,
    val sdkVersion: String
)

data class UserContext(
    val anonymousId: String,
    val userId: String?,
    val traits: Map<String, Any?>
)

data class SessionContext(
    val sessionId: String,
    val startedAt: Long
)

data class EventContext(
    val app: AppContext,
    val user: UserContext,
    val session: SessionContext,
    val consent: AnalyticsConsent
)

data class NormalizedEvent(
    val eventId: String,
    val eventName: String,
    val schema: String,
    val clientTime: Long,
    val properties: Map<String, Any?>,
    val context: EventContext
)

data class SimpleAnalyticsEvent(
    override val name: String,
    override val schema: String,
    override val properties: Map<String, Any?>
) : AnalyticsEvent

data class StayStartResult(
    val accepted: Boolean,
    val stayId: String? = null,
    val startedAt: Long? = null,
    val reason: String? = null
)

data class StayUpdateResult(
    val accepted: Boolean,
    val durationMs: Long? = null,
    val reason: String? = null
)

data class StayEndResult(
    val accepted: Boolean,
    val durationMs: Long? = null,
    val stayId: String? = null,
    val reason: String? = null
)

private data class StayState(
    val key: String,
    val stayId: String,
    val properties: Map<String, Any?>,
    val startedAt: Long,
    var activeStartedAt: Long,
    var accumulatedMs: Long = 0,
    var paused: Boolean = false,
    var pausedAt: Long? = null
)
